src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,328 @@
+                            continue;
+
+                        didBatch = true;
+                        // Send to client
+                        try
+                        {
+                            await WriteBatchToClientAndAck(batch, shard);
+                            batch.SendBatchToClientTcs.SetResult();
+                        }
+                        catch (Exception e)
+                        {
+                            // need to fail the shard subscription worker
+                            batch.SendBatchToClientTcs.TrySetException(e);
+                            throw;
+                        }
+                        await batch.ConfirmFromShardSubscriptionConnectionTcs.Task;
+
+                        // send confirm to client and continue processing
+                        await SendConfirmToClient();
+                        continue;
+                    }
+
+                    exceptions.Add(shard, shardHolder.PullingTask.Exception);
+
+                    if (stopping)
+                        continue;
+
+                    shardsToReconnect.Add(shard);
+
+                    if (shardHolder.LastErrorDateTime.HasValue == false)
+                    {
+                        shardHolder.LastErrorDateTime = DateTime.UtcNow;
+                        continue;
+                    }
+
+                    if (DateTime.UtcNow - shardHolder.LastErrorDateTime.Value <= _options.ShardedConnectionMaxErroneousPeriod)
+                        continue;
+
+                    // we are stopping this subscription
+                    stopping = true;
+                }
+
+                if (exceptions.Count == _shardWorkers.Count && stopping == false)
+                {
+                    stopping = true;
+                    foreach (var worker in _shardWorkers)
+                    {
+                        if (exceptions.TryGetValue(worker.Key, out var ex) == false)
+                            throw new InvalidOperationException($"Expected to get exception for shard worker {worker.Key} but could not find.");
+
+                        (bool shouldTryToReconnect, _) = worker.Value.Worker.CheckIfShouldReconnectWorker(ex, CancellationTokenSource, null, null, throwOnRedirectNodeNotFound: false);
+                        if (shouldTryToReconnect)
+                        {
+                            // we have at least one worker to try to reconnect
+                            stopping = false;
+                        }
+                    }
+                }
+
+                if (stopping)
+                {
+                    throw new AggregateException($"Stopping sharded subscription '{_options.SubscriptionName}' with id '{SubscriptionId}' for database '{TcpConnection.ShardedContext.DatabaseName}' because shard {string.Join(", ", exceptions.Keys)} workers failed.", exceptions.Values);
+                }
+
+                if (reconnectSp.ElapsedMilliseconds < 3000)
+                    continue;
```

Comment:
I don't like this, basically a busy loop with `sleep`

Reply:
done